test/test.py

```diff
@@ -0,0 +1,249 @@
+
+    def start(self):
+        pass
+
+    def sendToClient(self, data: bytes):
+        self.client.tcp.sendBytes(data)
+
+    def sendToServer(self, data: bytes):
+        self.server.tcp.sendBytes(data)
+
+    def recvFromClient(self, data: bytes):
+        self.client.tcp.recv(data)
+
+    def recvFromServer(self, data: bytes):
+        self.server.tcp.recv(data)
+
+    def setTimeStamp(self, timeStamp: float):
+        self.recorder.setTimeStamp(int(timeStamp * 1000))
+
+    def connectToServer(self):
+        pass
+
+    def startTLS(self):
+        pass
+
+    def sendPayload(self):
+        pass
+
+    def buildIOChannel(self, client, server):
+        """ Assert channel is built and mouse clicks and keyboard input """
+        self.builtIOChannel = True
+        super().buildIOChannel(client, server)
+
+        self.client.fastPath.addObserver(self.inputObserver)
+
+    def buildClipboardChannel(self, client, server):
+        """ Assert channel is built and add an observer to assert pasted content. """
+        self.builtCliprdrChannel = True
+        super().buildClipboardChannel(client, server)
+        clipboardMITM: PassiveClipboardStealer = self.channelMITMs[client.channelID]
+        self.clipboardObserver.addExpectedClipboardEntry("salutlagang\x00".encode("utf-16le"))
+        self.clipboardObserver.addExpectedClipboardEntry("lagang\x00".encode("utf-16le"))
+        clipboardMITM.client.addObserver(self.clipboardObserver)
+        clipboardMITM.server.addObserver(self.clipboardObserver)
+
+    def buildDeviceChannel(self, client, server):
+        self.builtRDPDRChannel = True
+        # not building the real channel as there is a bug in the test setup.
+        super().buildVirtualChannel(client, server)
+
+
+def main():
+    input_path = "test/files/test_session.pcap"
+    client_ip = "192.168.38.1"
+    mitm_ip = "192.168.38.1"
+    server_ip = "192.168.38.129"
+    output_path = "test/files/out/out.pyrdp"
+
+    logging.basicConfig(level=logging.CRITICAL)
+    logging.getLogger("scapy").setLevel(logging.ERROR)
+
+    packets = rdpcap(input_path)
+
+    test_mitm = TestMITM(output_path)
+
+    for packet in packets:
+        # The packets start with a Wireshark exported PDU structure
+        source, destination, destination_port, data = parseExportedPdu(packet)
+
+        try:
+            test_mitm.setTimeStamp(float(packet.time))
+            if source == client_ip and destination == mitm_ip and destination_port == 3389:
+                test_mitm.recvFromClient(data)
+            elif source == server_ip and destination == mitm_ip:
+                test_mitm.recvFromServer(data)
+            elif source == mitm_ip and destination == client_ip and destination_port != 3389:
+                test_mitm.sendToClient(data)
+            elif source == mitm_ip and destination == server_ip:
+                test_mitm.sendToServer(data)
+            else:
+                assert False
+        except NotImplementedError as e:
```

Comment:
Isn't this entire try-block redundant? Just let the exception be thrown, unless I'm missing something?

Reply:
Yeah, probably, it's from the pyrdp-replay script and I didnt remove it but probably should